<a href="https://colab.research.google.com/github/izzat-ai/learning-ai/blob/main/pandas/time_series.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

**Klinikaning kunlik Dataseti - AI dan olindi**

In [3]:
# Klinikaning 2026-yil boshidagi kunlik ma'lumotlar bazasi
data_ts = {
    'Sana': ['2026-01-01', '2026-01-02', '2026-01-03', '2026-01-05', '2026-01-06', '2026-01-07', '2026-01-08', '2026-01-09'],
    'Kelgan_Bemorlar': [45, 50, 38, 55, 62, 48, 58, 65],
    'Kunlik_Tushum': [12000000, 15000000, 9500000, 18000000, 21000000, 14000000, 19500000, 23000000] # ming so'mda
}

df_ts = pd.DataFrame(data_ts)
df_ts

,Sana,Kelgan_Bemorlar,Kunlik_Tushum
0,2026-01-01,45,12000000
1,2026-01-02,50,15000000
2,2026-01-03,38,9500000
3,2026-01-05,55,18000000
4,2026-01-06,62,21000000
5,2026-01-07,48,14000000
6,2026-01-08,58,19500000
7,2026-01-09,65,23000000


In [4]:
# DF haqida umumiy ma'lumot olish
df_ts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Sana             8 non-null      object
 1   Kelgan_Bemorlar  8 non-null      int64 
 2   Kunlik_Tushum    8 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 324.0+ bytes


- sana ustuni string ma'lumotlar turida ekan , uni datetime formatiga o'tkizish kerak

In [6]:
# sana ustunini datetime formatiga o'tkizish
df_ts['Sana'] = pd.to_datetime(df_ts['Sana'])
df_ts['Sana'].dtype

dtype('<M8[ns]')

In [7]:
df_ts

,Sana,Kelgan_Bemorlar,Kunlik_Tushum
0,2026-01-01,45,12000000
1,2026-01-02,50,15000000
2,2026-01-03,38,9500000
3,2026-01-05,55,18000000
4,2026-01-06,62,21000000
5,2026-01-07,48,14000000
6,2026-01-08,58,19500000
7,2026-01-09,65,23000000


In [9]:
# bemorlar kelgan sanalarni indeks qilish
df_ts = df_ts.set_index('Sana')
df_ts

,Kelgan_Bemorlar,Kunlik_Tushum
Sana,,
2026-01-01,45,12000000
2026-01-02,50,15000000
2026-01-03,38,9500000
2026-01-05,55,18000000
2026-01-06,62,21000000
2026-01-07,48,14000000
2026-01-08,58,19500000
2026-01-09,65,23000000


- sanalarga e'tibor bersak , 4-yanvar tushib qolgan

In [11]:
# 4-yanvarni indekslarga qo'shish
df_ts = df_ts.asfreq('D', fill_value=0)
df_ts

,Kelgan_Bemorlar,Kunlik_Tushum
Sana,,
2026-01-01,45,12000000
2026-01-02,50,15000000
2026-01-03,38,9500000
2026-01-04,0,0
2026-01-05,55,18000000
2026-01-06,62,21000000
2026-01-07,48,14000000
2026-01-08,58,19500000
2026-01-09,65,23000000


In [13]:
# haftalik hisobot olish
df_ts.resample('W').sum()

,Kelgan_Bemorlar,Kunlik_Tushum
Sana,,
2026-01-04,133,36500000
2026-01-11,288,95500000


- Yanvar oyining birinchi haftasida klinikaga 133 ta bemor kelgan va klinika 36.5 million so'm foyda ko'rgan
- huddi shu kabi ikkinchi haftada esa klinikaga kelgan bemorlar soni keskin oshgan va 95.5 million so'm foyda ko'rgan

In [16]:
# 3 kunlik siljidigan ma'lumotlarni aniqlash
bemorlar_3d = df_ts['Kelgan_Bemorlar'].rolling(3).mean()
bemorlar_3d

,Kelgan_Bemorlar
Sana,
2026-01-01,NaN
2026-01-02,NaN
2026-01-03,44.333333
2026-01-04,29.333333
2026-01-05,31.000000
2026-01-06,39.000000
2026-01-07,55.000000
2026-01-08,56.000000
2026-01-09,57.000000


- 1-va 2-yanvarlarda ma'lumot yetarli bo'lmagan shuning uchun ma'lumot yo'q
- 3-yanvargacha (3 kun ichida) klinikaga o'rtacha 44 ta bemor kelgan . Bu ko'rsatkich 4-yanvarga borib sezirarli pastlagan . Lekin 8-10-yanvarlar atrofi kelgan odamlar 56-57 tani tashkil qilgan